sdk/python/tests/compiler/performance_tests.py

```diff
@@ -0,0 +1,260 @@
+#!/usr/bin/env python3
+
+# Copyright 2021 kubeflow.org
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import concurrent.futures
+import datetime
+import functools
+import os
+import sys  # noqa
+import tempfile
+import time
+import threading
+
+from datetime import datetime as dt
+from os import environ as env
+from typing import Mapping, Callable
+
+from kfp_server_api import ApiRun, ApiRunDetail, ApiException
+from kfp_tekton.compiler import TektonCompiler
+from kfp_tekton._client import TektonClient
+from kfp_tekton.compiler.pipeline_utils import TektonPipelineConf
+
+
+# =============================================================================
+#  load test settings from environment variables
+# =============================================================================
+
+# TODO: turn env vars into script parameters
+PUBLIC_IP = env.get("PUBLIC_IP")
+NAMESPACE = env.get("NAMESPACE", None)
+USER_INFO = env.get("USER_INFO")
+CONNECT_SID = env.get("CONNECT_SID")
+NUM_WORKERS = int(env.get("NUM_WORKERS", 1))
+EXPERIMENT = env.get("EXPERIMENT_NAME", "PERF_TEST")
+OUTPUT_FILE = env.get("OUTPUT_FILE", f"perf_test_{dt.now().strftime('%Y%m%d_%H%M%S')}_{PUBLIC_IP}.csv")
+OUTPUT_SEP = env.get("OUTPUT_SEP", ",")
+
+
+print(f"Environment variables:\n\n"
+      f"  PUBLIC_IP:   {PUBLIC_IP}\n"
+      f"  NAMESPACE:   {NAMESPACE}\n"
+      f"  USER_INFO:   {USER_INFO}\n"
+      f"  CONNECT_SID: {CONNECT_SID}\n"
+      f"  NUM_WORKERS: {NUM_WORKERS}\n"
+      f"  EXPERIMENT:  {EXPERIMENT}\n"
+      f"  OUTPUT_FILE: {OUTPUT_FILE}\n"
+      f"  OUTPUT_SEP:  {OUTPUT_SEP}\n")
+
+
+# =============================================================================
+#  local settings that are not loaded from env variables
+# =============================================================================
+
+# kfp_tekton_root_dir = os.path.abspath(__file__).replace("sdk/python/tests/compiler/performance_tests.py", "")
+
+
+def get_client() -> TektonClient:
+    host = f"http://{PUBLIC_IP}/pipeline"
+    cookies = f"connect.sid={CONNECT_SID}; userinfo={USER_INFO}" if CONNECT_SID and USER_INFO else None
+    client = TektonClient(host=host, cookies=cookies)
+    client.set_user_namespace(NAMESPACE)  # overwrite system default with None if necessary
+    return client
+
+
+# method annotation to ensure the wrapped function is executed synchronously
+def synchronized(function):
+    lock = threading.Lock()
```

Comment:
could this lock be moved down to inside the wrapped function to even narrow down the locking scope? or you put it here to create the lock above the wrapped function on purposed

Reply:
the only function that needs to be synchronized, in the current implementation of this script is the pipeline compilation

Reply:
I will have to rewrite this script to record execution times differently in order to record compile time accurately, but the compilation times are below a 1/10 of a second so I don't think it matters to much

Reply:
I rewrote the test script such that compile times are recorded accurately now, despite having to synchronize access to the `compile_pipeline` method.